rgw: fetch_remote_obj() need to re-encrypt sse-s3 encrypted objects

clwluvw opened the pull request:
As SSE-S3 encrypted objects are tied to the source bucket's encryption key, they must be re-encrypted using the destination bucket's key to eliminate any dependencies on the source bucket's key.
The same principle applies to replication.

Fixes: https://tracker.ceph.com/issues/70446

clwluvw committed:
rgw/crypt: no context bounding to requesting user

requesting user can be a non-bucket owner, so picking the tenant can
result in different contexts. Picking bucket owner ensures the
consistent context.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
rgw: weaning off fetch_bucket_key_id() from req_state

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
rgw: weaning off get_sse_s3_bucket_key() from req_state

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
rgw: weaning off expand_key_name() from req_state

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
rgw: weaning off make_canonical_context() from req_state

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
rgw: always decrypt AES256 objects when skip_decrypt

As sse-s3 encryption is bind to the bucket's key, we should always
serve the objects decrypted so it can be reencrypted again by the
destination bucket's key id.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
rgw: extract handle sse-s3 encryption as a function

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
qa/rgw: add sse-s3 replication test

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -3618,6 +3627,29 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
+
+      std::unique_ptr<BlockCrypt> block_crypt;
+      std::string err_msg;
+      ret = handle_sse_s3_encryption(dpp, cct, y, bucket.get(), obj.key, &err_msg, src_attrs, &block_crypt, nullptr);
```
Experiencing a deadlock here while it tries to create a key for the destination bucket here: https://github.com/ceph/ceph/blob/db43ab3df9421eac7676101b0fc65e6e20b9fb8e/src/rgw/rgw_kms.cc#L309

clwluvw replied:
Thanks @cbodley for finding the issue.

clwluvw committed:
rgw: allow http receive and handle header to pause

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw committed:
rgw: regenerate sse-s3 keys on fetch_remote_obj()

When the source object is encrypted via sse-s3, fetch_remote_obj()
need to pause the receive and generate a new sse-s3 key for the
destination bucket and object and continue to receive and process
filter based on the new enc filter.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -3668,13 +3679,55 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
+  }
+
+  void make_sse_s3_key(bool *pause) override {
+    if (pause) *pause = true; // xxx: could this race with the spawn? (i.e. pause is set to curl after the spawn)
```
could this race with the spawn? (i.e. pause is set to curl after the spawn)

cbodley replied:
looks like it, yeah. `cancel()` is probably okay since it's just using `curl_multi_remove_handle()` and not interacting with the pause stuff. but `unpause_receive()` could potentially race

instead of just passing a `bool *pause` to the callback, maybe we need to pass enough state that the callback can lock and pause itself. then `make_sse_s3_key()` could safely to do that before calling `spawn()` to avoid the race with cancel/unpause

clwluvw replied:
wouldn't this by nature again fall into the same deadlock I've experienced (in sending HTTP request to Vault) in my first implementation if the current request was not paused?

cbodley replied:
if this is using the `rgw_http_req_data::mutex`, the vault request would create a different `rgw_http_req_data`

cbodley replied:
sorry, i didn't read `RGWHTTPClient::receive_http_header()` carefully enough. we're already holding that mutex over the call to `req_data->client->receive_header()`, and `req_data->read_paused = true` gets assigned under the same lock

so i'm not sure it _could_ race with the coroutine's call to `RGWHTTPStreamRWRequest::unpause_receive()`. that has to acquire the same mutex, so it would block waiting for `RGWHTTPClient::receive_http_header()` to finish

clwluvw replied:
right it's the same `req_data->lock` when we read headers and when we want to unpause. so i guess that could never race. thanks for the hint.

clwluvw commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -3668,13 +3679,55 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
+
+      in_stream_req->unpause_receive();
+    }, [] (std::exception_ptr eptr) {
+      if (eptr) std::rethrow_exception(eptr);
```
I guess here we should only log?

cbodley replied:
the functions called by this coroutine return error codes instead of throwing. if we only expect to get exceptions if something catastrophic happens (like bad_alloc), then i think it's best to rethrow them

cbodley commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -700,7 +700,18 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
 {
   if (skip_decrypt) { // bypass decryption for multisite sync requests
-    return 0;
+    auto crypt_mode = attrs.find(RGW_ATTR_CRYPT_MODE);
+    if (crypt_mode != attrs.end()) {
+      if (crypt_mode->second.to_str() != "AES256") {
+        // AES256 needs to be decrypted so that the target zone can re-encrypt with destination bucket's key
```
with rgw's default replication for DR, the source bucket and destination bucket are always the same, where i assume we'd never need to re-encrypt the objects. can we find a way to make this specific to bucket replication policy?

cbodley replied:
it's hard to make changes here that preserve compatibility between ceph versions

maybe we should leave the server side unchanged (always skip decrypt), and rely on `fetch_remote_obj()` to fetch both source and destination bucket keys so it can decrypt/reencrypt the body itself when necessary

multisite really shouldn't transfer objects in plaintext when they're supposed to be encrypted on both sites

clwluvw replied:
> it's hard to make changes here that preserve compatibility between ceph versions

With the introduction of `skip-enc-sse-s3`, if the destination requests it from a source that hasn't been upgraded yet, the object will only be retrieved in plain text if it's actually encrypted with SSE-S3 (that's where we don't send `skip-encryption` and only send `skip-enc-sse-s3` which the source doesn't understand it yet) and that is fine because the destination will create the keys and encrypt it again. Otherwise, the destination will send `skip-encryption` anyway, and the object will remain encrypted. The only scenario where this might not work is when both the source and destination are the same, and we don't want to decrypt it. However, this would only impact performance until the upgrade is complete, which I believe is negligible.

Vice-versa, the old version always requests `skip-encryption`, and the upgraded version will always respect that.

> multisite really shouldn't transfer objects in plaintext when they're supposed to be encrypted on both sites

Is this due to any security constraints? if so isn't (enforcing) https enough for that?

cbodley replied:
> > multisite really shouldn't transfer objects in plaintext when they're supposed to be encrypted on both sites
> 
> Is this due to any security constraints? if so isn't (enforcing) https enough for that?

tls helps, but what does enforcement mean? failing bucket replication for sse-s3-encrypted objects, i guess?

clwluvw replied:
> > > multisite really shouldn't transfer objects in plaintext when they're supposed to be encrypted on both sites
> > 
> > 
> > Is this due to any security constraints? if so isn't (enforcing) https enough for that?
> 
> tls helps, but what does enforcement mean? failing bucket replication for sse-s3-encrypted objects, i guess?

Yeah as this might be a config error by the admin, I guess this can be logged as an ERROR so the admin can make the call or ignore it via a config.

cbodley replied:
i'm just not sure why we'd want to introduce an error case that could be avoided by handling decrypt/reencrypt in `fetch_remote_obj()`. is that more difficult that i'm imagining it?

clwluvw replied:
>> maybe we should leave the server side unchanged (always skip decrypt), and rely on fetch_remote_obj() to fetch both source and destination bucket keys so it can decrypt/reencrypt the body itself when necessary
>
> handling decrypt/reencrypt in fetch_remote_obj(). is that more difficult that i'm imagining it?

hmm, i find this a bit inefficient to decrypt and encrypt on the destination zone at the same time, i thought if we could leave the decryption to the source zone why not? or do you think this is negligible (I'm assuming at some point this https://github.com/ceph/ceph/pull/52697 would be utilized and everything would fall in this case).

clwluvw replied:
> client traffic isn't effected

i wasn't referring only to client traffic but the replication time and eventually the replication performance at scale.

> i'm not sure it makes much difference?

But I'm not sure. if we don't have the right way to do it at least the other way, i guess the best is to start with the one you suggested and measure the performance see if this is really bothering so we improve later (even with not a clean way) with a reason.

clwluvw replied:
but to decrypt and reencrypt perhaps we would need some implementations from https://github.com/ceph/ceph/pull/54543

cbodley replied:
> hmm, i find this a bit inefficient to decrypt and encrypt on the destination zone at the same time, i thought if we could leave the decryption to the source zone why not?

sorry, i thought we were talking about the cpu overhead of decryption. we either need to pay that cost on the server side or the client side, and either will count towards replication latency. am i missing something?

clwluvw replied:
no that is true, just thought if we distribute the load (decryption to source and encryption to destination) that could leave the benefit of better resource utilisation than gathering all at one place.

clwluvw replied:
> it's hard to make changes here that preserve compatibility between ceph versions

One another approach i could think of:
What if we base the decision to apply the encryption filter on a response header, such as `x-rgw-decrypted`?
This would allow us to determine whether the source zone we're interacting with has been upgraded and whether the content has actually been decrypted. If it has, we could then apply the encryption filter accordingly. If not, the sync would maintain its current behavior, ensuring backward compatibility.
we would need to introduce `skip-decrypt-sse-s3` anyway to hint source not to if it was encrypted via sse-s3 when it's between buckets.

clwluvw committed:
rgw: introduce decrypt-mode to allow sse-s3 decryption in multisite

Passing decrypt-mode rather than a bool skip-decrypt would allow the
dest zone to pass "skip-except-sse-s3" when we have the src and dest
bucket tre not he same and passing "skip" when they are.
so with that, source zone can make the right call based on the string
provided by the decrypt-mode param.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

soumyakoduri commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -1029,6 +1029,11 @@ int RGWHTTPStreamRWRequest::handle_header(const string& name, const string& val,
   }
+
+  if (name == "X_AMZ_SERVER_SIDE_ENCRYPTION" && val == "AES256") {
+    cb->make_sse_s3_key(pause);
```
This results in crash while transitioning objects to cloud.  `cb` can be nullptr ( in case of HEAD requests) and needs to be checked here.

soumyakoduri commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -3813,13 +3825,55 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
+      std::unique_ptr<BlockCrypt> block_crypt;
+      std::string err_msg;
+      ret = handle_sse_s3_encryption(dpp, cct, yield, bucket.get(), obj.key, &err_msg, enc_attrs, &block_crypt, nullptr);
+      if (ret < 0) {
```
These changes affect cloud-restore and copy-object functionality too. 

In case of cloud-restore, the RGW bucket to which the object is being restored to from the cloud may not have encryption enabled. In such case we need to skip these re-encyption steps and just continue the download, instead of failing the op. Similar issue may arise with copy-object across the zones if target bucket doesn't have encryption enabled.

In addition, we may need to account for the cases where in destination bucket may have different encryption (not ss3_s3) enabled.
